source/iterator.go

```diff
@@ -0,0 +1,280 @@
+// Copyright © 2023 Meroxa, Inc. & Yalantis
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package source
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/data/azcosmos"
+	"github.com/conduitio-labs/conduit-connector-cosmos-nosql/common"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/huandu/go-sqlbuilder"
+)
+
+const (
+	allColumns          = "*"
+	table               = "c"
+	limit               = "OFFSET 0 LIMIT %d"
+	field               = "%s.%s"
+	lessThanOrEqual     = "%s <= %s"
+	greaterThan         = "%s > %s"
+	lastProcessedParam  = "@lastProcessed"
+	latestSnapshotParam = "@latestSnapshot"
+)
+
+// iterator is an implementation of an iterator for Azure Cosmos DB for NoSQL.
+type iterator struct {
+	position        *position
+	containerClient *azcosmos.ContainerClient
+	partitionKey    azcosmos.PartitionKey
+
+	container   string
+	keys        []string
+	orderingKey string
+	batchSize   uint
+
+	items []sdk.StructuredData
+}
+
+// newIterator creates a new instance of the iterator.
+func newIterator(ctx context.Context, config Config, sdkPosition sdk.Position) (*iterator, error) {
+	iter := &iterator{
+		partitionKey: azcosmos.NewPartitionKeyString(config.PartitionValue),
+		container:    config.Container,
+		keys:         config.Keys,
+		orderingKey:  config.OrderingKey,
+		batchSize:    config.BatchSize,
+	}
+
+	if len(iter.keys) == 0 {
+		iter.keys = []string{iter.orderingKey}
+	}
+
+	cred, err := azcosmos.NewKeyCredential(config.PrimaryKey)
```

Comment:
Could you please explain more what we do here

Reply:
Done 👍